### PR TITLE
Update Checkpoint Documentation

### PR DESCRIPTION
**Pull request by tgroh**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

Checkpoint finalization is best effort. A checkpoint that is committed
to durable state is permitted to be reused to start a reader, regardless
of if it is finalized.

Note that checkpoints which have an affect on their source (via
finalize) should generally require Deduping, due to the potential for
arbitrary checkpoint finalization failures.

### CODE REVIEW

**Comment by tgroh** [comment]
R: @dhalperi 

CC: @dpmills

**Comment by dhalperi** [comment]
R: @dpmills 

Get an LGTM from @dpmills and then I'll take a look -- thanks!

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java [hunk 1]
```diff
@@ -230,8 +228,11 @@ public boolean requiresDeduping() {
      * <p>All elements read up until this method is called will be processed together as a bundle.
      * (An element is considered 'read' if it could be returned by a call to {@link #getCurrent}.)
      * Once the result of processing those elements and the returned checkpoint have been durably
-     * committed, {@link CheckpointMark#finalizeCheckpoint} will (eventually) be called on the
-     * returned {@link CheckpointMark} object.
+     * committed, {@link CheckpointMark#finalizeCheckpoint} will be called at most once at some
+     * later point on the returned {@link CheckpointMark} object. Checkpoint finalization is
+     * best-effort, and checkpoints may not be finalized. If duplicate elements may be produced if
+     * checkpoints are not finalized in a timely manner, {@link UnboundedSource#requiresDeduping()}
+     * should be overridden.
```
Comment (dpmills) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java]: And getRecordId should be overridden to provide unique IDs

Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java]: Done.

**Pull request closed without merging**